corehq/apps/hqwebapp/utils.py

```diff
@@ -81,3 +86,91 @@ def aliased_language_name(lang_code):
             if code == lang_code:
                 return name
         raise KeyError('Unknown language code %s' % lang_code)
+
+
+def extract_password(password):
+    # Passwords set with expected salts length and padding would respect this regex
+    reg_exp = r"^sha256\$([a-z|0-9|A-Z]{6})(.*)([a-z|0-9|A-Z]{6})=$"
+    match_result = re.match(reg_exp, password)
+    # strip out outer level padding of salts/keys and ensure three matches
+    if match_result and len(match_result.groups()) == 3:
+        match_groups = re.match(reg_exp, password).groups()
+        hash_left = match_groups[0]
+        hash_right = match_groups[2]
+        stripped_password = match_groups[1]
+        # decode the stripped password to get internal block
+        # decoded(salt1 + encoded_password + salt2)
+        try:
+            decoded_password = base64.b64decode(stripped_password)
+        except TypeError:
+            return ''
+        match_result_2 = re.match(reg_exp, decoded_password)
+        # strip out hashes from the internal block and ensure 3 matches
+        if match_result_2 and len(match_result_2.groups()) == 3:
+            match_groups = match_result_2.groups()
+            # ensure the same hashes were used in the internal block as the outer
+            if match_groups[0] == hash_left and match_groups[2] == hash_right:
+                # decode to get the real password
+                password_hash = re.match(reg_exp, decoded_password).groups()[1]
+                # return password decoded for UTF-8 support
+                try:
+                    return base64.b64decode(password_hash).decode('utf-8')
+                except TypeError:
+                    return ''
+            else:
+                # this sounds like someone tried to hash something but failed so ignore the password submitted
+                # completely
+                return ''
+        else:
+            # this sounds like someone tried to hash something but failed so ignore the password submitted
+            # completely
+            return ''
+    else:
+        # return the password received AS-IS
+        return password
+
+
+@memoized
+def get_password_hasher():
+    return get_hasher()
+
+
+def hash_password(password):
+    hasher = get_password_hasher()
+    return hasher.encode(password, hasher.salt())
+
+
+def verify_password(password, password_salt):
+    hasher = get_password_hasher()
+    return hasher.verify(password, password_salt)
+
+
+# quickcache for multiple decode attempts in the same request:
+# 1. an attempt to decode a password should be done just once in a request for the login attempt
+# check to work correctly.
+# 2. there should be no need to decode a password multiple times in the same request.
+@quickcache(['password_hash'], timeout=0)
```

Comment:
Thank You @millerdev @dannyroberts 

```
You'll need to do it somewhere where you have access to the request object itself or cache the result on an object with the same lifecycle as the request (for example, the view).
```
i don't think it would be possible to track down all views calling that function. It might also be within django at places.

I looked at the documentation for `quickcache` and seemed like it was the right way. Added that it should vary on password_hash argument passed and by default it expires in 10 seconds. Also added timeout 0 so that its not saved in any other cache.
```
In addition to caching in the default shared cache,
          quickcache caches in memory for 10 seconds
          (conceptually the length of a single request).
```